bolt/lib/Rewrite/RewriteInstance.cpp

```diff
@@ -2439,6 +2439,26 @@ void RewriteInstance::readDynamicRelocations(const SectionRef &Section,
     if (Symbol)
       SymbolIndex[Symbol] = getRelocationSymbol(InputFile, Rel);
 
+    const uint64_t ReferencedAddress = SymbolAddress + Addend;
+    BinaryFunction *Func =
+        BC->getBinaryFunctionContainingAddress(ReferencedAddress);
```

Comment:
We need to find a function that contains `SymbolAddress` and then verify that it also contains `ReferencedAddress`.

Reply:
I think this check might be unnecessary. While testing this patch, I found that `R_*_RELATIVE` relocations are of the form `*ABS*+Addend`. IE:
```
$ objdump -DR /usr/bin/* | grep 'R_AARCH64_RELATIVE*
<addrs1>: R_AARCH64_RELATIVE       *ABS*+0xaac0
<addrs2>: R_AARCH64_RELATIVE       *ABS*+0xb20a4
```
I have added a check to make sure we exit if that is not the case.

Reply:
Cool. If `R_*_RELATIVE` is the only relocation type you want to handle here, that should work. The dynamic linker will ignore they symbol even if it's present.

The question remains what happens if we have dynamic relocations of non-relative type against text. We should handle it even if we don't anticipate it to happen in practice.

Reply:
>The question remains what happens if we have dynamic relocations of non-relative type against text

In that situation this change has no effect and we can revert back to the current status quo: using `--skip-funcs` to deal with errors.
>We should handle it even if we don't anticipate it to happen in practice.

Since this patch doesn't introduce errors in the case of dynamic relocations of non-relative type, I think we should handle those cases in a separate patch.

Reply:
Fair enough.